plugins/workspace-plugin/tests/workspace-folder-updater.spec.ts

```diff
@@ -34,149 +35,283 @@ Object.assign(theia, { workspace });
 
-    workspaceFolderUpdater.addWorkspaceFolder(PROJECT_1_PATH);
-    workspaceFolderUpdater.addWorkspaceFolder(PROJECT_2_PATH);
-    await workspaceFolderUpdater.addWorkspaceFolder(PROJECT_2_PATH);
-    await waitUpdateFolders(500);
+    await workspaceFolderUpdater.addWorkspaceFolder(PROJECT_1);
+    await workspaceFolderUpdater.addWorkspaceFolder(PROJECT_2);
+    await workspaceFolderUpdater.addWorkspaceFolder(PROJECT_2);
 
-    expect(actualWorkspaceFolders).toStrictEqual([PROJECT_1_PATH, PROJECT_2_PATH]);
-    expect(updateWorkspaceFolders).toBeCalledTimes(2);
+    expect(actualWorkspaceFolders).toStrictEqual([PROJECT_1, PROJECT_2]);
+    expect(updateWorkspaceFoldersMock).toBeCalledTimes(2);
     expect(dispose).toBeCalledTimes(2);
   });
 
-  test('it shoud skip adding Workspace Folders duplicates', async () => {
-    const workspaceFolder = { uri: uriStub(PROJECT_1_PATH) };
-    addedFoldersForEvent = [workspaceFolder];
+  test('Adding of Workspace Folder should be rejected by timeout', async () => {
+    skipWorkspaceFolderUpdate = true;
+    try {
+      await workspaceFolderUpdater.addWorkspaceFolder(PROJECT_1);
+    } catch (error) {
+      expect(error.message).toEqual(`Adding of workspace folder ${PROJECT_1} was canceled by timeout`);
+      return;
+    }
+
+    fail();
+  });
+
+  test('Reject adding the first folder, sucessful adding of the next one', async () => {
+    // add PROJECT_1_PATH
+    // add PROJECT_2_PATH
+    // case:
+    //    adding of PROJECT_1_PATH must be failes
+    //    PROJECT_2_PATH must be added to workspace folders
 
-    await workspaceFolderUpdater.addWorkspaceFolder(PROJECT_1_PATH);
+    allowedFoldersToAdd = [PROJECT_2];
 
-    theia.workspace.workspaceFolders = [{ name: '', index: 0, uri: workspaceFolder.uri }];
+    let addingProject1Failed = false;
+    try {
+      await workspaceFolderUpdater.addWorkspaceFolder(PROJECT_1);
+    } catch (e) {
+      expect(e.message).toMatch(`Unable to add workspace folder ${PROJECT_1}`);
+      addingProject1Failed = true;
+    }
 
-    await workspaceFolderUpdater.addWorkspaceFolder(PROJECT_1_PATH);
-    await waitUpdateFolders(150);
+    expect(addingProject1Failed).toBe(true);
 
-    expect(actualWorkspaceFolders).toStrictEqual([PROJECT_1_PATH]);
-    expect(updateWorkspaceFolders).toBeCalledTimes(1);
-    expect(dispose).toBeCalledTimes(1);
+    await workspaceFolderUpdater.addWorkspaceFolder(PROJECT_2);
+
+    expect(actualWorkspaceFolders).toStrictEqual([PROJECT_2]);
+    expect(updateWorkspaceFoldersMock).toBeCalledTimes(2);
+    expect(dispose).toBeCalledTimes(2);
+  });
+
+  test('Workspace Folders should be added in proper sequence', async () => {
+    allowedFoldersToAdd = [PROJECT_1, PROJECT_2, PROJECT_3];
+
+    await workspaceFolderUpdater.addWorkspaceFolder(PROJECT_1);
+    await workspaceFolderUpdater.addWorkspaceFolder(PROJECT_2);
+    await workspaceFolderUpdater.addWorkspaceFolder(PROJECT_3);
```

Comment:
We shouldn't rely on the condition that the next request for adding/removing a folder comes when the previous request has been completed.
Moreover: the responsibility of the `WorkspaceFolderUpdater` entity is management of adding and removing projects as workspace folders with respect [the documentation](https://github.com/eclipse-theia/theia/blob/8b9fb43c763f2f1d179f9839ef993d7218d99073/packages/plugin/src/theia.d.ts#L5902).  

The previous implementation took into account that and the tests covered the corresponding use cases, [here](https://github.com/eclipse-che/che-theia/blob/1782bb6e932e4cd9a369f9551846f3cbef5dd0ce/plugins/workspace-plugin/tests/workspace-folder-updater.spec.ts#L129-L142), for example. 

I believe you have greatly improved `WorkspaceFolderUpdater` entity and there are should be no problem.
 
So, please, just remove `await` for the first and the second call of `addWorkspaceFolder()` method. I guess, for the third call we should have `await` to check the state when all projects are added as workspace folders.

Reply:
I will not change anything in the test.
And the following call of `addWorkspaceFolder` or `removeWorkspaceFolder` will always wait for the previous promise to be resolved. Just because they are placed in a chain.
But ok, removing them for all tests. Just to see that implementation works properly.

Reply:
> the following call of addWorkspaceFolder or removeWorkspaceFolder will always wait for the previous promise to be resolved. Just because they are placed in a chain.

I believe it's true at the moment and it's the details of the implementation - we should pin that logic in the tests.
Otherwise, let's say, in 1 month, someone change the implementation and the tests will sill be happy, but the real use case is broken

Reply:
Extra `await` have been removed. I will push it together with tests improvement.